foundations/javascript_basics/understanding_errors.md

```diff
@@ -0,0 +1,132 @@
+  return c;
+}
+
+function print() {
+  add();
+}
+
+print();
+```
+
+Our function `print()` should call on `add()`, which returns a variable named `c`, which currently has not been declared. The corresponding error is as follows:
+
+<img width="620" alt="Screen Shot 2021-05-17 at 12 15 06 PM" src="https://user-images.githubusercontent.com/22967723/118543770-90641980-b709-11eb-8c85-4f49195901bc.png">
+
+The stack trace tells us that:
+
+1. `c is not defined` in scope of `add()`, which is declared on line 5
+2. `add()` was called by `print()`, which was declared on line 9
+3. `print()` itself was called on line 12.
+   The stack trace allows you to follow the execution of functions to find the code responsible for throwing the error .
+
+# Common Types of Errors
+
+These are some of the most common errors you will encounter, so it’s important to understand them.
+
+## Syntax Error
+
+A syntax error occurs when the code you are trying to run is syntactically incorrect. For example this:
+
+```js
+function helloWorld() {
+console.log “Hello World!”
+}
+```
+
+will throw the following error, because we forgot the parenthesis for `console.log()`!
+<img width="602" alt="Screen Shot 2021-05-17 at 12 15 11 PM" src="https://user-images.githubusercontent.com/22967723/118543788-9823be00-b709-11eb-8f2a-3e4f64cee880.png">
+[MDN | SyntaxError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SyntaxError)
+
+## Reference Error
+
+We covered reference errors in the first example in this lesson, but it’s important to remember that these arise because whatever variable you are trying to reference does not exist.
+
+[MDN | ReferenceError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ReferenceError)
+
+## Type Error
+
+These errors are thrown for a few different reasons:
+
+Per MDN, a `TypeError` may be thrown when:
+
+> - an operand or argument passed to a function is incompatible with the type expected by that operator or function;
+> - or when attempting to modify a value that cannot be changed;
+> - or when attempting to use a value in an inappropriate way.
+
+Say we have two strings that you would like to combine to create one message such as below:
+
+```js
+const str1 = "Hello";
+const str2 = "World!";
+const message = str1.push(str2);
+```
+<img width="621" alt="Screen Shot 2021-05-17 at 12 14 12 PM" src="https://user-images.githubusercontent.com/22967723/118543650-6579c580-b709-11eb-981d-4bdb74966ba4.png">
+
+Here, we get a `TypeError` with a message stating that `str1.push is not a function`. This is a common error message that confuses learners, because we know that `.push()` is certainly a function! You’ve probably used it to add items to _arrays_ before, but that’s the key- `.push()` is not a String method, it’s an Array method. Hence, it is “not a function” that you can find as a String method. If we change `.push()` to `.concat()`, a proper String method, our code runs as intended!
```

Comment:
Could help to add a note here to have people make sure they know what type they're trying to run operations against. Keeping data type in mind will help, this happens to be a good place to remind them.